src/dmd/target.d

```diff
@@ -586,10 +587,11 @@ struct Target
     /**
      * Params:
      *   tf = function type to check
+     *   fd = declaration of function to check (null if unavailable, i.e. just checking type)
      * Returns:
      *   true if return value from function is on the stack
      */
-    extern (C++) static bool isReturnOnStack(TypeFunction tf)
+    extern (C++) static bool isReturnOnStack(TypeFunction tf, FuncDeclaration fd)
```

Comment:
This doesn't seem right; taking both a `TypeFunction` and a `FunctionDeclaration`.  I don't have a solution at the moment, but my intuition tells me there must be a better way.

Reply:
Doesn't really feel very offensive to me. I've seen lots of functions over the years follow this pattern for the same reasons.

Reply:
We have gone through this above https://github.com/dlang/dmd/pull/8330#issuecomment-394592022. I would like a better solution as well.

Reply:
> fd.isMember2() && !fd.isStatic()

Just noticed that there is `fd.needThis()` for this check.

Reply:
> `(fd && Target.isReturnOnStack(fd)) || (!fd && Target.isReturnOnStack(tf))`

That looks very idiomatic-DMD, actually.  I would prefer an even better solution, but if I had to choose between that and the current proposal, I'd go with that.

Reply:
If we use overloads, I would replace `retStyle()` with `Target.returnsOnStack()` to avoid having 4 similar functions.

Reply:
Sounds good to me.  I don't see why we need both `retStyle` and `returnsOnStack`.